### Hypothesis Testing

In [1]:
import statsmodels.api as sm
import numpy as np
import pandas as pd

#### One Population Proportion

Research Question:

In previous years, 52% of parents believed that electronics and social media was a cause of their teenager's lack of sleep. Do more parents today believe that their teenager's lack of sleep is caused due to electronics and social media?

Population: Parents with teenagers (age 13 - 18)

Parameter of interest: p

Null Hypothesis: p = 0.52

Alternative Hypothesis: p > 0.52

1018 parents

56% believe that their teenager's lack of sleep is caused due to electronics and social media.

In [2]:
n = 1018
pnull = 0.52 # null hypothesis
phat = 0.56  # alternative hypothesis
sm.stats.proportions_ztest(phat * n, n, pnull)

# phat = observed proportion of parents
# n = amount of parents that participated in the study
# n (2nd) = population size

(2.571067795759113, 0.010138547731721065)

The result shows that:

Test statistic output (alpha value) = 2.57

p-value = 0.01

p-value < alpha value (@95% level of confidence)

Hence, reject null hypothesis.

Population proportion of parents who believe that their teenager's lack of sleep is caused due to electronics and social media is greater than 52%

#### Difference in Population Proportions

Research Question:

Is there a significant difference between the population proportions of parents of black children and parents of Hispanic children who report that their child has had some swimming lessons?

Population: All parents of black children age 6-18 and all parents of Hispanic children age 6-18

Parameter of interest: p1 - p2, where p1 = black and p2 = Hispanic

Null Hypothesis: p1 - p2 = 0

Alternative Hypothesis: p1 - p2 is not equal to 0

247 Parents of Black children 36.8% of parents report that their child has had some swimming lessons.

308 Parents of Hispanic children 38.9% of parents report that their child has had some swimming lessons.

In [3]:
n1 = 247 #parents of black children that participated in the study
p1 = 0.37 #proportion of parents

n2 = 308 # parents of Hispanic children that participated in the study
p2 = 0.39 # proportion of parents

population1 = np.random.binomial(1, p1, n1)
population2 = np.random.binomial(1, p2, n2)
population1[:5]

array([0, 0, 0, 1, 0])

In [4]:
population2[:5]

array([1, 1, 1, 0, 0])

Result: 
For p1 => only 1 of the first 5 parents had reported that their child has had a swimming lessons.

For p2 => only 1 of the first 5 parents had also reported that their child has had a swimming lessons but in a different location of the population array.

In [5]:
# This example implements the analysis from the 'Difference in Two Proportion' Lecture videos

# sampe sizes
n1 = 247 
n2 = 308 

# Number of parents reporting that their child has had swimming lessons.
y1 = 91 
y2 = 120

# Estimates of the population proportions
p1 = round(y1 / n1, 2)
p1 = round(y2 / n2, 2)

# Estimates of the combined population proportions
phat = (y1 + y2) / (n1 + n2) 

sm.stats.ttest_ind(population1, population2)

(-1.7671314138006304, 0.07775772629496062, 553.0)

The result shows that:

Test statistic output (alpha value) = -1.56

p-value = 0.11

p-value > alpha value (@95% level of confidence)

Hence, we do not reject the null hypothesis.

#### One Population Mean

Research Question:

Is the average cartwheel distance (in inches) for adults more than 80 inches?

Population: All adults

Parameter of interest: population mean cartwheel distance. 

Null hypothesis: pop. mean = 80

Alternative hypothesis: pop. mean > 80

25 Adults

pop. mean = 82.46

std = 15.06

In [6]:
df = pd.read_csv("C:\\Users\\user\\Documents\\Jupyter notebooks for Data Science\\Cartwheeldata.csv")

In [7]:
df.head()

ID  Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1   56      F            1       Y             1    62.0      61.0   
1   2   26      F            1       Y             1    62.0      60.0   
2   3   33      F            1       Y             1    66.0      64.0   
3   4   39      F            1       N             0    64.0      63.0   
4   5   27      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y              1      7  
1          70        Y              1      8  
2          85        Y              1      7  
3          87        Y              1     10  
4          72        N              0      4

In [8]:
n = len(df)
mean = df["CWDistance"].mean()
sd = df["CWDistance"].std()
(n, mean, sd)

(25, 82.48, 15.058552387264855)

In [9]:
sm.stats.ztest(df["CWDistance"], value = 80, alternative = "larger") #indicating an alternative hypothesis where mean is larger than null hypothesis

(0.8234523266982029, 0.20512540845395266)

The result shows that:

Test statistic output (mean value) = 0.82 (as against 0.80)

p-value = 0.2

p-value is pretty high

Hence, we cannot not reject the null hypothesis.

#### Difference in Population Means

Research Question:

Considering adults in the NHANES data, do males have a sinificantly higher mean Body Mass Index than females?

Population: Adults in the NHANES data

Parameter of interest: mean1 - mean2, Body Mass Index. 

Null hypothesis: mean1 = mean2

Alternative hypothesis: mean1 not equal to mean2

2976 Females mean = 29.94

std1 = 7.75

2759 Male Mexican-American Adults mean2 = 28.78

std2 = 6.25

mean1 - mean2 = 1.16

In [10]:
url = ("C:\\Users\\user\\Documents\\Jupyter notebooks for Data Science\\nhanes_2015_2016.csv")

In [11]:
da = pd.read_csv(url)
da.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
3       1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4       1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  
3     37.7     38.3     110.1     2.0  
4     36.0     27.2      80.4     2.0  

[5 rows x 28 columns]

In [12]:
females = da[da["RIAGENDR"] == 2]
male = da[da["RIAGENDR"] == 1]

In [13]:
females.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
3   83735     2.0     1.0     1.0       2         2        56         3   
4   83736     2.0     1.0     1.0       2         2        42         4   
5   83737     2.0     2.0     NaN       2         2        72         1   
7   83742     1.0     NaN     1.0       2         2        32         1   
12  83752     1.0     NaN     2.0       1         2        30         2   

    DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
3        1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4        1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   
5        2.0       2.0  ...   122.0    58.0   64.4  150.0    28.6    34.4   
7        2.0       4.0  ...   114.0    70.0   64.5  151.3    28.2    34.1   
12       1.0       4.0  ...   104.0    50.0   71.2  163.6    26.6    37.3   

    BMXARML  BMXARMC  BMXWAIST  HIQ210  
3      37.7     38.3     110.1     2.0  
4      36.0     27.2      80.4     2.0  
5      33.5     31.4      92.9     NaN  
7      33.1     31.5      93.3     2.0  
12     35.7     31.0      90.7     2.0  

[5 rows x 28 columns]

In [14]:
male.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
6  83741     1.0     NaN     8.0       1         1        22         4   
8  83743     NaN     NaN     NaN       2         1        18         5   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
6       1.0       4.0  ...   112.0    74.0   76.6  165.4    28.0    38.8   
8       1.0       NaN  ...     NaN     NaN   72.4  166.1    26.2     NaN   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  
6     38.0     34.0      86.6     NaN  
8      NaN      NaN       NaN     2.0  

[5 rows x 28 columns]

In [15]:
n1 = len(females)
mu1 = females["BMXBMI"].mean()
sd1 = females["BMXBMI"].std()

(n1, mu1, sd1)

(2976, 29.939945652173996, 7.75331880954568)

In [16]:
n2 = len(male)
mu2 = male["BMXBMI"].mean()
sd2 = male["BMXBMI"].std()

(n2, mu2, sd2)

(2759, 28.778072111846985, 6.252567616801485)

In [17]:
sm.stats.ztest(females["BMXBMI"].dropna(), male["BMXBMI"].dropna())

(6.1755933531383205, 6.591544431126401e-10)

Result: 

The test statistic is large while the p-value is small.

We reject the null hypothesis that the BMI for the 2 populations is the same.

### Walk-Through_ Hypothesis Testing with NHANES

#### Hypothesis Testing

In this notebook, we demonstrate formal hypothesis testing using the NHANES data.

It is important to note that the NHANES data are a "complex survey". The data are not an independent and representative sample from the target population. Proper analysis of complex survey data should make use of additional information about how the data were collected. Since complex survey analysis is a somewhat specialized topic, we ignore this aspect of the data here, and analyze the NHANES data as if it were an independent and identically distributed sample from a population.

In [18]:
# import the necessary libraries
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg') #workaround, there may be a better way
import seaborn as sns
%matplotlib inline
import matplotlib.pyplot as plt
import statsmodels.api as sm
import scipy.stats.distributions as dist

Below we read the data, and convert some of the integer codes to text values

In [19]:
url = ("C:\\Users\\user\\Documents\\Jupyter notebooks for Data Science\\nhanes_2015_2016.csv")

In [20]:
da = pd.read_csv(url)

da['SMQ020x'] = da.SMQ020.replace({1: 'Yes', 2: 'No', 7: np.nan, 9: np.nan})
da['RIAGENDRx'] = da.RIAGENDR.replace({1: 'Male', 2: 'Female'})
da['DMDCITZNx'] = da.DMDCITZN.replace({1: 'Yes', 2: 'No', 7: np.nan, 9: np.nan})

In [21]:
da['SMQ020x'].head()

0    Yes
1    Yes
2    Yes
3     No
4     No
Name: SMQ020x, dtype: object

In [22]:
da['RIAGENDRx'] = da.RIAGENDR.replace({1: 'Male', 2: 'Female'})
da['RIAGENDRx'].head()

0      Male
1      Male
2      Male
3    Female
4    Female
Name: RIAGENDRx, dtype: object

### Hypothesis tests for one proportions

The most basic hypothesis test may be the one-sample test for a proportion. This test is used if we have specified a particular value as the null value for the proportion, and we wish to assess if the data are compatible with the true parameter value being equal to this specified value. One-sample tests are not very often in practice, because it is not very common that we have a specific fixed value to use for comparison. For illustration, imagine that the rate of lifetime smoking in the US were different from 40%. In the following notebook cell, we carry out the (two-sided) one-sample test that the population proportion of smokers is 0.4 and obtain a p-value of 0.43. This indicates that the NHANES data are compatible with the proportion of (ever) smokers in the US being 40%.  

In [23]:
x = da.SMQ020x.dropna() =='Yes' # population of smokers
p = x.mean() # the mean population proportion
se = np.sqrt(0.4 * 0.6 / len(x)) # standard error

In [24]:
test_stat = (p - 0.4) / se # test statistic
test_stat

0.7823563854332805

In [25]:
pvalue = 2*dist.norm.cdf(-np.abs(test_stat)) # the p-value/absolute value of our test statistic
print(test_stat, pvalue)

0.7823563854332805 0.4340051581348052


Result:

Test statistic is very small indicating that our population proportion is not really far from our null hypothesis.

While the p-value is extremely high.

We do not reject the null hypothesis here. So, we continue to believe that the population proportion of smokers is 40%

The following cell carries out the same test as performed above using the Statsmodels library. The rresults in the first (default) case below are slightly different from the resuts obtained above because Statsmodels by default uses the sample proportion instead of the null proportion when computing the standard error. The distinction is rarely consequential, but we can specify that the null proportion should be used to calculate the standar error and the results agree exactly with what we calculated above. The first two lines below uses the exact binomil sampling distribution. We can see here that the p-values are nearly identical in all three cases. This is expected when the sample size is large, and the proportion is not close to either 0 or 1.

In [26]:
print(sm.stats.proportions_ztest(x.sum(), len(x), 0.4))
print(sm.stats.proportions_ztest(x.sum(), len(x), 0.4, prop_var = 0.4))
print(sm.stats.binom_test(x.sum(), len(x), 0.4)) # Exact binomial p-value

(0.7807518954896244, 0.43494843171868214)
(0.7823563854332805, 0.4340051581348052)
0.4340360854410036


C:\Users\user\anaconda3\lib\site-packages\scipy\stats\_discrete_distns.py:67: RuntimeWarning: divide by zero encountered in _binom_pdf
  return _boost._binom_pdf(x, n, p)


### Hypothesis tests for two proportions

Comparative


In [27]:
dx = da[['SMQ020x', 'RIDAGEYR', 'RIAGENDRx']].dropna() #Drop missing values
dx = dx.loc[(dx.RIDAGEYR >= 20) & (dx.RIDAGEYR <= 25), :] # Restrict to peope between 20 & 25

# summarize the data by calculating the proportion of yes responses and the sample size
p = dx.groupby('RIAGENDRx')['SMQ020x'].agg([lambda z: np.mean(z == 'Yes'), 'size'])
p.columns = ['Smoke', 'N']
print(p)

              Smoke    N
RIAGENDRx               
Female     0.238971  272
Male       0.341270  252


In [28]:
# The pooled rate of yes responses, and the standard error of the estimated difference of proportions
p_comb = (dx.SMQ020x == 'Yes').mean()
va = p_comb * (1 - p_comb)
se = np.sqrt(va * (1 / p.N.Female + 1 / p.N.Male))

In [29]:
# Calculate the test statistic and its p-value
test_stat = (p.Smoke.Female - p.Smoke.Male) / se
pvalue = 2 * dist.norm.cdf(-np.abs(test_stat))
print(test_stat, pvalue)

# test statistic is extremely small and p-value is also small. We reject null hyp.
# the population proportion of female and male smokers is actually different

-2.5833303066279414 0.009785159057508375


Essentially the same test as above can be conducted by converting the "Yes/No" responses to numbers (Yes = 1, No = 0) and conducting a two_sample t-test, as below:

In [30]:
dx_females = dx.loc[dx.RIAGENDRx=='Female', 'SMQ020x'].replace({'Yes': 1, 'No': 0})
dx_males = dx.loc[dx.RIAGENDRx=='Male', 'SMQ020x'].replace({'Yes': 1, 'No': 0})
sm.stats.ttest_ind(dx_females, dx_males) # prints test statistics, p-value, degree of freedom

(-2.5949731446269344, 0.00972590232121254, 522.0)

### Hypothesis tests comparing means



In [32]:
dx = da[['BPXSY1', 'RIDAGEYR', 'RIAGENDRx']].dropna() 
dx = dx.loc[(dx.RIDAGEYR >= 40) & (dx.RIDAGEYR <= 50) & (dx.RIAGENDRx =='Male'), :]
print(dx.BPXSY1.mean()) # prints mean blood pressure
sm.stats.ztest(dx.BPXSY1, value =120) # prints test statistic, p-value



125.86698337292161


(7.469764137102597, 8.033869113167905e-14)

In the cel below, we carry out a formal test of the null hypothesis that the mean blood pressure for women between the ages of 50 and 60 is equal to the mean blood pressure of 

In [34]:
dx = da[['BPXSY1', 'RIDAGEYR', 'RIAGENDRx']].dropna() 
dx = dx.loc[(dx.RIDAGEYR >= 50) & (dx.RIDAGEYR <= 60), :]
bpx_female = dx.loc[dx.RIAGENDRx =='Female', 'BPXSY1']
bpx_male = dx.loc[dx.RIAGENDRx =='Male', 'BPXSY1']
print(bpx_female.mean(), bpx_male.mean()) # prints female mean, male mean 
print(sm.stats.ztest(bpx_female, bpx_male)) # prints test statistic, p-value
print(sm.stats.ttest_ind(bpx_female, bpx_male)) # prints test statistic, p-value, degrees of freedom

127.92561983471074 129.23829787234044
(-1.105435895556249, 0.2689707570859362)
(-1.105435895556249, 0.26925004137768577, 952.0)


There is a high p-value. So we cannot reject the null hypothesis that these two populations have different symbolic blood pressure P.

Another important aspect of two-sample mean testing is heteroscedasticity', meaning that the variances within the two groups being compared may be different. 

In [35]:
dx = da[['BPXSY1', 'RIDAGEYR', 'RIAGENDRx']].dropna()
da['agegrp'] = pd.cut(da.RIDAGEYR, [18, 30, 40, 50, 60, 70, 80])
da.groupby(['agegrp', 'RIAGENDRx']) ['BPXSY1'].agg(np.std).unstack()

RIAGENDRx     Female       Male
agegrp                         
(18, 30]   10.467543  10.529873
(30, 40]   12.392896  13.097883
(40, 50]   15.410072  16.470465
(50, 60]   18.580881  18.369981
(60, 70]   18.529034  18.945549
(70, 80]   20.801830  20.747324

The standard error of the mean difference (e.g. mean female blood pressure minus mean male blood pressure) can be estimated in at least two

In [39]:
for k, v in da.groupby('agegrp'):
    bmi_female = v.loc[v.RIAGENDRx == 'Female', 'BMXBMI'].dropna()
    bmi_female = sm.stats.DescrStatsW(bmi_female)
    bmi_male = v.loc[v.RIAGENDRx == 'Male', 'BMXBMI'].dropna()
    bmi_male = sm.stats.DescrStatsW(bmi_male)
    print(k)
    print('pooled: ', sm.stats.CompareMeans(bmi_female, bmi_male).ztest_ind(usevar = 'pooled'))
    print('unequal: ', sm.stats.CompareMeans(bmi_female, bmi_male).ztest_ind(usevar = 'unequal'))
    print()                

(18, 30]
pooled:  (1.7026932933643306, 0.08862548061449803)
unequal:  (1.7174610823927183, 0.08589495934713169)

(30, 40]
pooled:  (1.4378280405644919, 0.15048285114648174)
unequal:  (1.4437869620833497, 0.1487989105789246)

(40, 50]
pooled:  (2.8933761158070186, 0.003811246059501354)
unequal:  (2.9678691663536725, 0.0029987194174035366)

(50, 60]
pooled:  (3.362108779981383, 0.0007734964571391287)
unequal:  (3.3754943901739387, 0.0007368319423226156)

(60, 70]
pooled:  (3.617240144243268, 0.00029776102103194453)
unequal:  (3.628483094544553, 0.00028509141471492935)

(70, 80]
pooled:  (2.926729252512241, 0.003425469414486057)
unequal:  (2.9377798867692064, 0.0033057163315194853)

